🔄 refactor: MCP Registry System with Distributed Caching

nhtruong opened the pull request:
## Summary

This PR restructures the MCP registry system by breaking down the monolithic `MCPServersRegistry` into specialized, purpose-driven components. The registry now serves as the single source of truth for all MCP server configurations across the application, with a clear separation of concerns and improved maintainability.

## Key Changes

### 1. Registry Restructure - Separation of Concerns

The original `MCPServersRegistry` has been split into three specialized components:

- **`MCPServersRegistry`** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:18-81`)
  - Simplified to focus solely on managing and storing MCP server configurations
  - Acts as the **single source of truth** for all server configs in the application
  - Provides unified interface for retrieving configs with proper fallback hierarchy
  - No longer handles initialization or inspection logic

- **`MCPServerInspector`** (`packages/api/src/mcp/registry/MCPServerInspector.ts`)
  - Handles server inspection, health checks, and capability discovery
  - Detects OAuth requirements and fetches server metadata
  - Converts server tools to LibreChat-compatible format

- **`MCPServersInitializer`** (`packages/api/src/mcp/registry/MCPServersInitializer.ts`)
  - Manages server initialization logic with distributed leader-follower coordination
  - Supports multiple deployment types: single-node (in-memory) and multi-node (Redis-backed)
  - Fault-tolerant design handles leader crash scenarios gracefully
  - Prevents cross-deployment contamination using in-memory initialization flag in combination with distributed status cache
  - Handles timeout mechanisms for unresponsive servers

### 2. Three-Tier Server Configuration System

The registry now manages three distinct categories of server configurations:

1. **Shared App Servers** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:19`)
   - Auto-started servers available to all users
   - Initialized at application startup
   - Loaded from `librechat.yaml`
   - Now can be modified on the fly

2. **Shared User Servers** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:20`)
   - User-scope servers requiring OAuth or other forms of authentication
   - Loaded from `librechat.yaml`
   - Now can be modified on the fly

3. **Private User Servers** (`packages/api/src/mcp/registry/MCPServersRegistry.ts:21`) - **NEW**
   - Per-user configurations dynamically added during runtime
   - Not loaded from config files - created programmatically
   - Fully dynamic - can be added/removed on the fly per user
   - Isolated per user ID with dedicated cache instances

### 3. Simplified Access Patterns

**Before**: MCPManager had numerous wrapper methods for accessing server configs
```typescript
// Multiple methods wrapping registry access
getAllServers(), getSharedServers(), getUserServers(), etc.
```

**After**: Direct access to registry singleton
```typescript
// Direct access - MCPManager now accesses registry directly
const configs = await registry.getAllServerConfigs(userId);
const config = await registry.getServerConfig(serverName, userId);
```

**Benefits:**
- Cleaner, more maintainable code
- Reduced indirection and complexity
- Consistent access patterns across the codebase
- Registry truly becomes the source of truth

### 4. Distributed Caching Layer

Added a comprehensive caching system for server configurations:

- **`ServerConfigsCacheRedis`**: Redis-backed configuration cache for distributed environments
- **`ServerConfigsCacheInMemory`**: In-memory fallback cache for single-node deployments
- **`RegistryStatusCache`**: Manages shared metadata and statuses for the registry across nodes.
- **`ServerConfigsCacheFactory`**: Factory pattern for cache creation with in-memory/redis cache support

### 5. Improved Utilities

- **`withTimeout`** (`packages/api/src/utils/promise.ts`): Replaces `Promise.race` patterns with clearer timeout handling
- Added comprehensive integration tests for all cache implementations
- Added GitHub Actions workflow for cache integration tests

## Technical Details

### Distributed Initialization Process

The initialization system (`packages/api/src/mcp/registry/MCPServersInitializer.ts:38-63`) is designed to handle various deployment scenarios with fault tolerance and contamination prevention:

#### Deployment Type Support

**Single-Node Deployments (In-Memory)**
- Uses in-memory cache (`ServerConfigsCacheInMemory`) for server configurations
- No Redis dependency required
- Simpler, faster initialization for standalone instances

**Multi-Node Deployments (Redis-Backed)**
- Uses Redis cache (`ServerConfigsCacheRedis`) for distributed state management
- Leader-follower coordination via `RegistryStatusCache`
- Only one node (leader) performs expensive initialization operations
- Follower nodes wait and poll until leader signals completion

#### Fault Tolerance

**Leader Crash Handling**
- If the leader node crashes during initialization, followers detect the failure
- After a 3-second delay, all followers independently attempt initialization
- The first follower to become leader completes the initialization process
- No manual intervention required - system self-heals automatically

**Initialization State Management**
- Leader resets caches and performs server inspection/initialization
- Upon completion, leader signals via `statusCache.setInitialized(true)`
- Followers poll `statusCache.isInitialized()` and proceed once ready
- Ensures only one node performs expensive operations even under failure conditions

#### Cross-Deployment Contamination Prevention

**Per-Deployment Initialization**
- Uses **in-memory** `initialized` flag (not cache-based) at `MCPServersInitializer.ts:22`
- Each deployment maintains its own initialization state independently
- Prevents scenarios where one deployment's cached state affects another deployment
- Critical for blue-green deployments, rolling updates, and multi-tenant environments

**Design Rationale** (`packages/api/src/mcp/registry/MCPServersInitializer.ts:28-29`):
> Uses in-memory `initialized` flag (not cache-based) to prevent cross-deployment contamination.
> Each deployment must initialize its own instance to ensure configs are loaded correctly.

This ensures that even if Redis contains stale initialization state from a previous deployment, each new deployment will properly initialize its own registry with fresh configurations.


---

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor 
- [x] This change requires a documentation update

## Testing

- Comprehensive integration tests against actual Redis servers.
- Separate test sets for both in-memory and Redis deployments.

## Checklist

Please delete any irrelevant options.

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] Any changes dependent on mine have been merged and published in downstream modules.

nhtruong committed:
refactor: Restructure MCP registry system with caching

- Split MCPServersRegistry into modular components:
  - MCPServerInspector: handles server inspection and health checks
  - MCPServersInitializer: manages server initialization logic
  - MCPServersRegistry: simplified registry coordination
- Add distributed caching layer:
  - ServerConfigsCacheRedis: Redis-backed configuration cache
  - ServerConfigsCacheInMemory: in-memory fallback cache
  - RegistryStatusCache: distributed leader election state
- Add promise utilities (withTimeout) replacing Promise.race patterns
- Add comprehensive cache integration tests for all cache implementations
- Remove unused MCPManager.getAllToolFunctions method

leblancfg commented on .github/workflows/cache-integration-tests.yml:
```diff
@@ -8,12 +8,14 @@ on:
 jobs:
   cache_integration_tests:
-    name: Run Cache Integration Tests
+    name: Integration Tests that use actual Redis Cache
```
```suggestion
    name: Integration Tests using the Redis Cache
```

atefbellaaj-at-slalom commented on packages/api/src/mcp/registry/__tests__/mcpConnectionsMock.helper.ts:
consider adding .helper. in the ignored pattern, so it will not fail the test:ci
example:
```
"test:ci": "jest --coverage --ci --testPathIgnorePatterns=\"\\.*integration\\.|\\.*helper\\.\"",
```

danny-avila committed:
fix: Update OAuth flow to include user-specific headers

danny-avila committed:
chore: Update Jest configuration to ignore additional test files

- Added patterns to ignore files ending with .helper.ts and .helper.d.ts in testPathIgnorePatterns for cleaner test runs.

danny-avila committed:
fix: oauth headers in callback

danny-avila committed:
chore: Update Jest testPathIgnorePatterns to exclude helper files

- Modified testPathIgnorePatterns in package.json to ignore files ending with .helper.ts and .helper.d.ts for cleaner test execution.

danny-avila committed:
ci: update test mocks
